### PR TITLE
WIP: Firefox policies w/ extensions integration

### PR DESCRIPTION
**Pull request by eyJhb**
### Description
Ability to provide a list of extensions, that will be installed in a correct way (none-hacky), which is supported from the Mozilla point of view.

Currently this will not work, as it also requires some work to be done on how the Firefox addons are generated.

This also relies on the work here https://github.com/NixOS/nixpkgs/pull/94898 , as a custom patch is required to make the policies on a pr. profile basis.
It can be made user wide, but that would not be that useful for some.

Maybe supporting both would be good? Any input @rycee , on what you think?

This is the config I use atm.

```
{ config, pkgs, lib, ... }:

let
  ffxpis = pkgs.recurseIntoAttrs (pkgs.callPackage ./modules/firefox {});
in {
    programs.firefox = {
      enable = true;
      package = pkgs.firefox-policies;

      profiles = {
        "default" = {
          id = 0;
          extensions = with ffxpis; [
            https-everywhere
            bitwarden
            i-dont-care-about-cookies
            ublock-origin
            surfingkeys_ff
          ];


          settings = {
            "toolkit.legacyUserProfileCustomizations.stylesheets" = true;
            "svg.context-properties.content.enabled" = true;
          };
          extraPolicies = {
            DontCheckDefaultBrowser = false;
            OfferToSaveLogins = false;
          };

          userChrome = ''
          '';
        };
      };
    };
}
```

Where I use this version, to generate my plugins. Slightly modified default.nix, with a python rewrite of the current version - https://gitlab.com/eyJhb/nix-firefox-addons

### Checklist

- [ ] Change is backwards compatible.

- [ ] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/rycee/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/rycee/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/rycee/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/rycee/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by rycee** [comment]
Hmm, I don't have any particular opinion. I image it would be nice to have a per-profile setup. The extensions are now installed per-profile although the setting is global, I've been thinking about moving the option to be per-profile.

I don't understand how it could work with the `XDG_RUNTIME_DIR`, though. Since that directory gets deleted regularly and the HM activation only happens occasionally you might not actually start Firefox while the file is available? `XDG_RUNTIME_DIR` seems like an odd choice for this file in general, it doesn't seem ephemeral and non-essential.

**Comment by eyJhb** [comment]
The current way is somewhat hacky compared to this, as it is a more official way of doing it (using policies, that is).

I have however not thought about `/run/user/.../` being deleted that often. The reason for placing it there, is that it is the place where Firefox will read the policies from, if the patch in the Nixpkgs is not applied (which allows for the pr. Profile settings/extensions). Not sure if there is a better way.

Ideally we would get the nixpkgs PR merged, butI can also see forcing users to use that specific Firefox with that patch is not ideal. But so is forcing any Firefox version/package on a user isn't. So I think that supporting both a normal FF version and a patched one is ideal. Or is that just me?

Is there a way to make the `/run/user/.../`, file always available using HM? I am guessing the only way would be a systemd unit, or patching the package like in the other PR that is open.

**Comment by eyJhb** [comment]
Thinking I will integrate it with a  systemd path unit, which will ensure that the file exists :) Not sure it will be before Friday/Saturday. Any comments is appreciated !

**Comment by rycee** [comment]
Should be mentioned that a systemd unit would make the module incompatible with macos. But I imagine the use of `XDG_RUNTIME_DIR` already is causing an incompatibility 😕

I'm wondering if not this policy file is intended for something completely different than actually setting up Firefox in a general way? The way its done using `XDG_RUNTIME_DIR` seems to very focused on some use-case that doesn't match ours. Perhaps it's more for when you'd like to run a containerized Firefox?

**Comment by eyJhb** [comment]
Slip of a finger regarding the closing.

It is actually called enterprise policies, so it is a way for a company to configure it in a special way. Normally it would reside in`/etc/firefox/policies.json`, until someone complained about it, and got it enabled for using pr. User. The directory chosen does not seem optimal no. But there should be a similar place for Mac OS X I would suppose?

**Comment by rycee** [comment]
Yeah, I imagine it should be available for macos as well but I don't really know anything about it :-/ If it's not available then I fear this would have to go in a separate module `firefox-policies.nix`, or something, which is only enabled for Linux hosts.

**Comment by eyJhb** [comment]
We might have to pull in a Mac user then, and see how that works. That could be a option yes, and would be quite nice if that is the end result.

I guess that would have to be cleared first. However, how do you feel about the nixpkgs patch?

**Comment by rycee** [comment]
I don't really know anything about the Firefox code to say whether the patch itself is fine or not but in general I imagine it would be quite nice to have per-profile policies 🙂

But I worry a bit about requiring the user to use a special firefox-policies package. It might be worth doing a first iteration of this feature in HM without per-profile policies and then bring in the per-profile support later?

**Comment by eyJhb** [comment]
The code itself is a thing, but it was more the general idea of having it :)

Hmm, that might be a good thing yes. Let me try and see, if I can clean this PR up a little bit. I will be making the changes in the assumption, that it will work on Mac too, and that it can replace the current way extensions work. If that is not the case, it should be easy to put it back in again :)

Need to figure out how systemd wants to cooperate with a path unit, but I am not sure that can be a solution at all, sadly... I think the best choice it a timer script, that runs ever X secs to check if the file exists, if not symlink it :)

**Comment by rycee** [comment]
Since the runtime dir is created when you log in I imagine it should be sufficient with a oneshot service wanted by `default.target`.

**Comment by eyJhb** [comment]
Update - Not forgotten, but have had (and still have) a lot of preparations for family gathering.
Hoping to work on this soon again!

**Comment by eyJhb** [comment]
@rycee I have made it into a oneshot service, but it does not seem to run the service on `nixos-rebuild switch`, or am I doing something wrong?

**Comment by eyJhb** [comment]
Seems to work with `systemd.user.startServices = true`, as the bash implementation did not work well. This should either be set for the user, or indicate it in a option.

Besides this, I am still debating if it should have the option to include the patch for Firefox, so that we can have a pr. profile policies.json. I really like the idea of it! Because you could have one for uni, hacking (accept all certs), work, personal each with its own extensions, bookmarks, etc.

Any comment is appreciated @rycee  !

File: modules/programs/firefox.nix [hunk 1]
```diff
@@ -304,16 +370,54 @@ in
...
+            Type = "oneshot";
+            RemainAfterExit = true;
+
+            ExecStart = let
+              policiesFile = pkgs.writeText "policies.json" (mkPolicies cfg);
+            in (toString (pkgs.writeShellScript "write-firefox-policies-file" ''
+              set -xe
+
+              # set our runtime dir + policies location here 
+              XDG_RUNTIME_DIR="/run/user/$($DRY_RUN_CMD  ${pkgs.coreutils}/bin/id -u)"
```
Comment (rycee) [inline modules/programs/firefox.nix:404]: `$DRY_RUN_CMD` won't be defined inside this shell script, it's only for activation scripts. So you can remove it.

Comment (eyJhb) [reply modules/programs/firefox.nix]: Fixed in newest `update` commit. Will squash them once everything is good-to-go :)

**Comment by eyJhb** [comment]
@rycee it should be working now, is there anything that should be changed? And can this even be merged? :)

**Comment by eyJhb** [comment]
Should maybe be reworked to work with #91724 instead :)

**Pull request closed without merging**